Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1918,66 +1948,161 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
     textTransform.ty = std::round(textTransform.ty);
 
     CGAffineTransform finalTextTransform = CGAffineTransformConcat(textTransform, deviceTransform);
-    if ((fabs(finalTextTransform.a * glyphRun->fontEmSize) <= c_glyphThreshold &&
-         fabs(finalTextTransform.b * glyphRun->fontEmSize) <= c_glyphThreshold) ||
-        (fabs(finalTextTransform.d * glyphRun->fontEmSize) <= c_glyphThreshold &&
-         fabs(finalTextTransform.c * glyphRun->fontEmSize) <= c_glyphThreshold)) {
-        TraceWarning(TAG, L"Glyphs too small to be rendered");
-
-        // Not a failure state! Not drawing the glyphs is *okay*.
-        return S_FALSE;
+    for (size_t i = 0; i < runsCount; ++i) {
+        if ((fabs(finalTextTransform.a * glyphRuns[i].run->fontEmSize) <= c_glyphThreshold &&
+             fabs(finalTextTransform.b * glyphRuns[i].run->fontEmSize) <= c_glyphThreshold) ||
+            (fabs(finalTextTransform.d * glyphRuns[i].run->fontEmSize) <= c_glyphThreshold &&
+             fabs(finalTextTransform.c * glyphRuns[i].run->fontEmSize) <= c_glyphThreshold)) {
+            TraceWarning(TAG, L"Glyphs too small to be rendered");
+
+            // Not a failure state! Not drawing the glyphs is *okay*.
+            return S_FALSE;
+        }
     }
 
-    auto& state = CurrentGState();
-
-    // If text is only flipped vertically, we can draw it all at once rather than glyph by glyph
-    if ((textTransform.a == 1.0f && fabs(textTransform.d) == 1.0f && textTransform.b == 0.0f && textTransform.c == 0.0f) ||
-        !transformByGlyph) {
-        RETURN_IF_FAILED(
-            Draw(_kCGCoordinateModeUserSpace, &textTransform, [&state, glyphRun](CGContextRef context, ID2D1DeviceContext* deviceContext) {
-                deviceContext->DrawGlyphRun(D2D1::Point2F(0, 0), glyphRun, state.fillBrush.Get(), DWRITE_MEASURING_MODE_NATURAL);
-                return S_OK;
-            }));
+    if (((textTransform.a == 1.0f && fabs(textTransform.d) == 1.0f && textTransform.b == 0.0f && textTransform.c == 0.0f) ||
+         !transformByGlyph) &&
+        state.textDrawingMode == kCGTextFill) {
+        // Able to fast-path simple text drawing that doesn't need to clip/stroke nor has text transformations
+        RETURN_IF_FAILED(Draw(_kCGCoordinateModeDeviceSpace, nullptr, [&](CGContextRef context, ID2D1DeviceContext* deviceContext) {
+            D2D1_POINT_2F origin{ 0, 0 };
+            for (size_t i = 0; i < runsCount; ++i) {
+                if (glyphRuns[i].attributes) {
+                    CGColorRef fontColor = (CGColorRef)CFDictionaryGetValue(glyphRuns[i].attributes, _kCGForegroundColorAttributeName);
+                    if (fontColor) {
+                        CGContextSetFillColorWithColor(context, fontColor);
```

Comment:
Should this manipulate global drawing state, or should it set the D2D brush to use later?